### ISSUE TITLE
[POC] linting with pylint

### ISSUE THREAD

**Issue by adebardo** [GlacioHack/xdem#550]
# Context

The goal of this ticket is to implement linting on the code. We have noticed that certain rules applied to demcompare are not applied in the files of xDEM. 

Examples of errors present in the test files:
- `tests/test_volume.py:252:11: W0125: Using a conditional statement with a constant value (using-constant-test)`
- `tests/test_spatialstats.py:563:12: W0612: Unused variable 'ny' (unused-variable)`
- `tests/test_spatialstats.py:509:0: C0115: Missing class docstring (missing-class-docstring)`
- `tests/test_spatialstats.py:547:4: R0914: Too many local variables (27/15) (too-many-locals)`
- `tests/test_vcrs.py:22:8: C0206: Consider iterating with .items() (consider-using-dict-items)`

We cannot guarantee that the same code quality is applied to all the code. Therefore, we propose to add the pylint tool to the code dependencies and correct the errors highlighted.

# Tasks

- [ ] Explain why pylint is in the "test" dependencies in the [setup](https://github.com/GlacioHack/xdem/blob/main/setup.cfg) file and see if it's the most appropriate place for it
- [ ] Add a .pylintrc file
    - [ ] Use the one from demcompare (Amaury, if you can take a look and see if you want to modify anything)
```diff
# Maximum number of characters on a single line.
- max-line-length=80
+ max-line-length=120
- # Maximum number of lines in a module.
- max-module-lines=700
```
- [ ] Add a pylint dependency in the .pre-commit-config.yaml file (same config as demcompare)
- [ ] Correct the errors returned
    - [ ] If possible
    - [ ] Otherwise, add the # ignore and list them in a document
- [ ] Add pylint to CI github actions

**Comment by duboise-cnes** [comment]
on the general linting, i agree to put something at the beginning to benefit from the quality wrappers for all code to be done afterwards. I think it is a must have with this too permissive python language. I would nevertheless prefer go to : 
- quicker ruff directly for replacement of pylint (not all but i think sufficient and growing), black, flake8, isort. only one tool easier
- and mypy very important also (type checking is really important). Maybe another ticket.

See https://docs.astral.sh/ruff/faq/ and ruff is now really mature and faster. So it would be a better choice on linting i would advise

"Explain why pylint is in the "test" dependencies in the [setup](https://github.com/GlacioHack/xdem/blob/main/setup.cfg) file and see if it's the most appropriate place for it" --> you mean use a "dev" target ? more explicit but if xdem is used with test for dev parts, go for it. No casus belli ;)

max line to 120 is a lot for my old me used of little screen but it is the dev standard so ok for me if the code is already there ;)

max-module-lines is also important for refactoring goal. 700 or 1000 max could be another issue in parallel of code structure possible refactoring

**Comment by adebardo** [comment]
Hello, I am glad you suggested using ruff. I will revise the ticket with this solution.

**Comment by adebardo** [comment]
Ruff seems bad as it job : 

with ruff :
![image](https://github.com/user-attachments/assets/efc7b613-8c3a-44e8-bb8a-307e81f01632)

with pylint: 
![image](https://github.com/user-attachments/assets/b20394df-9aab-43fc-a5a6-3a5cc87282a1)

**Comment by duboise-cnes** [comment]
Ok see internally if it is misconfiguration of ruff and if not enough mature (it's weird nonetheless) I think we keep isort, black, flake8, pylint, mypy as a baseline.

**Comment by adehecq** [comment]
I didn't know riff, but I think if we could reduce the number of linters to one that manages all, that would certainly make things easier.
It's seems normal that ruff does not complain in some parts while pylint does. As stated in the [FAQ](https://docs.astral.sh/ruff/faq/#how-does-ruffs-linter-compare-to-pylint), ruff and pylint do not completely overlap on rules. Maybe this can be tuned a bit with configuration, but otherwise we just have to accept that no tool will perfectly catch all issues.

As for the line length, 79 is really too short when there are 2, 3, 4 indentation levels... 120 is a bit more comfortable.
For the maximum number of lines, let's try to not be too strict for the moment, as we have rather long files in geoutils/xdem at the moment 😅

**Comment by rhugonnet** [comment]
Agree on all.
For `ruff`, I think @erikmannerfelt mentioned it about a year ago. At that time the main improvement seemed to be speed for very large projects that had to wait several minutes to perform `pre-commit`. This is not a problem on this repo yet, but it would be more durable to move to ruff anyways for sure, so all for it as well :slightly_smiling_face:

**Comment by adebardo** [comment]
I finally found a great config file for Ruff, and it seems suitable for our project. I didn’t ignore any rules, so we can discuss them during development if needed.

P.S.: We can also choose the line length.
